### PR TITLE
SI-7944 FOUND: stray undetermined type params in vicinity of implicits

### PR DESCRIPTION
**Pull request by retronym**
Implicit search created a nested Context into which the results of
its typechecking, namely, errors and undetermined type parameters
(roughly: those inferred as Nothing) are stashed.

The code the drives the process was checking for errors, but
discarded those undetermined type parameters.

This commit copies them from the child context to the parent,
which lets `Typer#adapt` to get to:

```
else if (hasUndetsInMonoMode) { // (9)
  assert(!context.inTypeConstructorAllowed, context) //@M
  instantiatePossiblyExpectingUnit(tree, mode, pt)
}
```

Our lost TypeVar has found its way home! The reward for which
is being instantiated, based on another type inference session
adapting the expression's type to the expected type.

### CODE REVIEW

**Comment by retronym** [comment]
Review by @adriaanm

**Comment by adriaanm** [comment]
And I didn't even have to go around and put up MISSING posters!

File: src/compiler/scala/tools/nsc/typechecker/Implicits.scala [hunk 1]
```diff
@@ -78,7 +78,11 @@ trait Implicits {
       })
       debuglog("update buffer: " + implicitSearchContext.reportBuffer.errors)
     }
-    context.undetparams = context.undetparams filterNot result.subst.from.contains
+    // SI-7944 undetermined type parameters that result from inference within typedImplicit land in
+    //         `implicitSearchContext.undetparams`, *not* in `context.undetparams`
+    //         Here, we copy them up to parent context (analogously to the way the errors are copied above),
+    //         and then filter out any which *were* inferred and are part of the substitutor in the implicit search result.
+    context.undetparams = ((context.undetparams ++ implicitSearchContext.undetparams) filterNot result.subst.from.contains).distinct
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Implicits.scala:85]: also of interest here: https://github.com/retronym/scala/blob/6fa9c78a56ab690de4db5bf7623f66545f2ac137/src/compiler/scala/tools/nsc/typechecker/Implicits.scala#L331

**Comment by retronym** [comment]
@cunei @luc @harrah 

Does this [failure](https://scala-webapps.epfl.ch/jenkins/job/pr-scala-integrate-ide/3665/console) look familiar?

```
[info] [error] /localhome/jenkins/b/workspace/pr-scala-integrate-ide/target/zinc/target-0.6.4-josh-hack-3/project-builds/zinc-77374c3a564c525ae9145677ec14c0bae74af3d3/src/main/scala/com/typesafe/zinc/SbtAnalysis.scala:58: value merge is not a member of object sbt.inc.Analysis
[info] [error]     val mergedAnalysis = Analysis.merge(analysesAndSetups map {_._1})
[info] [error]                                   ^
[info] [error] /localhome/jenkins/b/workspace/pr-scala-integrate-ide/target/zinc/target-0.6.4-josh-hack-3/project-builds/zinc-77374c3a564c525ae9145677ec14c0bae74af3d3/src/main/scala/com/typesafe/zinc/SbtAnalysis.scala:58: value merge is not a member of object sbt.inc.Analysis
[info] [error]     val mergedAnalysis = Analysis.merge(analysesAndSetups map {_._1})
```

The method reported as missing was recently added to SBT:

https://github.com/sbt/sbt/commit/30cdb1a

**Comment by gkossakowski** [comment]
Zinc master has been updated (typesafehub/zinc@54f56651020916e3c1425e7617488c472a653399) to use new apis introduced in sbt/sbt@30cdb1a but in PR validation zinc master is compiled against stable release of sbt.

I believe we should either compile stable versions of everything or masters of everything. I'd prefer to compile everything from stable versions (tags).

**Comment by gkossakowski** [comment]
Actually, it uses a bit old branch: https://github.com/typesafehub/sbt-builds-for-ide/blob/master/sbt-on-2.11.x#L56
I'll issue a PR which updates that branch to sbt 0.13.
Also, there'll be another PR which uses stable (0.3) version of zinc for building.

**Comment by gkossakowski** [comment]
However, I can do it only tomorrow's evening. Feel free to beat on this.

**Comment by adriaanm** [comment]
I'm working on the sbt build for #3087, so I'll take this. (Does the 0.13 branch build as-is on 2.11 these days?)

File: src/compiler/scala/tools/nsc/typechecker/Implicits.scala (hunk 1 above)
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Implicits.scala:85]: Thanks for the pointer. I traced this through to better understand the process.

By way of example, in:

``` scala
scala> val ls = List[Option[Any]]()
ls: List[Option[Any]] = List()

scala> ls.flatten
```

The undetermined type params from the outer context are `B(in method flatten)`, during the implicit search for the implicit parameter of type `Option[Any] => scala.collection.GenTraversableOnce[B]`.

That is first used to approximate the sought type as `Option[Any] => scala.collection.GenTraversableOnce[?]` (`wildPt`) which is used to prune the candidate implicits.

Later, when typechecking the plausible candidates, occurrences of the undet params are replaced by type variables in the expected type, and `matchesPt` is called to see if the candidate is suitable:

```
itree2 = {
  ((xo: Option[Any]) => scala.this.Option.option2Iterable[Any](xo))
}: Option[Any] => Iterable[Any]

matchesPt(
  tp = Option[Any] => Iterable[Any],
  pt = Option[Any] => scala.collection.GenTraversableOnce[?B]
  undet = [type B]
)
```

What does `matchesPt` do with the `undet`? Ultimately, they are used in `isApplicableToMethod`:

``` scala
      def typesCompatible(args: List[Type]) = undetparams match {
        case Nil => isCompatibleArgs(args, formals) && isWeaklyCompatible(mt resultType args, pt)
        case _   => tryInstantiating(args)
      }

      def tryInstantiating(args: List[Type]) = falseIfNoInstance {
        val restpe = mt resultType args
        val AdjustedTypeArgs.Undets(okparams, okargs, leftUndet) = methTypeArgs(undetparams, formals, restpe, args, pt)
        val restpeInst = restpe.instantiateTypeParams(okparams, okargs)
```

**Comment by retronym** [comment]
The latest force push just expands the test case a little.

**Comment by adriaanm** [comment]
LGTM

**Pull request merged**